### ISSUE TITLE
Regression in redirection functionality of System.Net.Http.HttpClient

### ISSUE THREAD

**Issue by ChengTian** [dotnet/corefx#1930]
@davidsh , I found this regression when trying to integrating `HttpClient` into `dnu restore`

Here is a simple repro:
```
public class Program
{
    public void Main(string[] args)
    {
        Task.Run(async () =>
        {
            var client = new HttpClient();
            //var request = new HttpRequestMessage(HttpMethod.Get, "http://www.bing.com");  // Works
            var request = new HttpRequestMessage(HttpMethod.Get, "http://bing.com");  // Throws exception
            using (var response = await client.SendAsync(request))
            {
                Console.WriteLine(await response.Content.ReadAsStringAsync());
            }
        }).Wait();
    }
}
```
The exception thrown:
```
System.Net.Http.HttpRequestException: An error occurred while sending the request. ---> System.Net.Http.WinHttpException: The requested header was not
 found

   at System.Net.Http.WinHttpHandler.WinHttpStatusCallback(IntPtr handle, IntPtr context, UInt32 internetStatus, IntPtr statusInformation, UInt32 stat
usInformationLength)
   --- End of inner exception stack trace ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter`1.GetResult()
   at ConsoleApp8.Program.<>c.<<Main>b__0_0>d.MoveNext() in C:\AspNet\VS2015Projects\ConsoleApp8\src\ConsoleApp8\Program.cs:line 19
System.Net.Http.WinHttpException: The requested header was not found

   at System.Net.Http.WinHttpHandler.WinHttpStatusCallback(IntPtr handle, IntPtr context, UInt32 internetStatus, IntPtr statusInformation, UInt32 stat
usInformationLength)
```

**Comment by davidsh** [comment]
Thanks for the bug report.  We can repro this.  The bug is happening in the sources that has not yet been posted to GitHub (WinHttpHandler).  We will fast-track a fix for this and get the MYGET feeds updated.

**Comment by davidsh** [comment]
cc: @weshaggard 

We fixed this today (internal issue #1183127).  It will probably be sometime Saturday until the MYGET feed gets updated with the next nightly build.

**Comment by SidharthNabar** [comment]
@ChengTian - Please give this fixed version of HttpClient a try and let us know if the bug still repros. Else we can close this issue. Thanks!

**Comment by ChengTian** [comment]
@SidharthNabar , the bug still repros with `System.Net.Http.WinHttpHandler 4.0.0-beta-23008`, which is the latest version I can find.

**Comment by davidsh** [comment]
4.0.0-beta-23008 does NOT yet have the fix in it.  We are waiting for another build of the binaries which should then have the fix in them.
I will post an update to this issue when I have confirmed that the binaries on the DNX/ASP.NET MYGET Dev feed have the fix in it.

**Comment by davidsh** [comment]
The fix for this bug is now in the 4.0.0-beta-23011 package for System.Net.Http.

This is available on the dotnet-core MYGET feed:
https://www.myget.org/gallery/dotnet-core

**Comment by davidsh** [comment]
cc: @ChengTian

**Comment by ChengTian** [comment]
Doesn't repro with fixed package. Closing this.

**Comment by Hochman85** [comment]
Hi I am actually still having the same weird issue with System.Net.Http.HttpRequestMessage when trying to add a custom header. The header is added and yet SendAsync throws an exception about "header was not found"! Here in the team, we have spent 4 days in vain to get this to work. Any help would be appreciated.

private static async Task MakeRequestExample(string url)
        {
            try
            {
                HttpMethod httpMethod = HttpMethod.Get;

                using (HttpClient httpClient = new HttpClient())
                {
                    using (HttpRequestMessage httpRequestMessage = new HttpRequestMessage(httpMethod, url))
                    {
                        //httpRequestMessage.Headers.Accept.TryParseAdd("text/plain"); <-- not sure about this             

                        bool added = httpRequestMessage.Headers.TryAddWithoutValidation("Client-Id", "314");
                        // added will be true and when I debug this I actually find the header

                        using (HttpResponseMessage httpResponse = await httpClient.SendAsync(httpRequestMessage))
                        {
                            switch (httpResponse.StatusCode)
                            {
                                case HttpStatusCode.OK:
                                    // logic 1
                                    break;

                                case HttpStatusCode.RedirectKeepVerb:
                                case HttpStatusCode.NotFound:
                                case HttpStatusCode.Moved:
                                case HttpStatusCode.BadRequest:
                                case HttpStatusCode.InternalServerError:
                                    // logic 2
                                    break;

                                default:
                                    // logic 3
                                    break;
                            }
                        }
                    }
                }
            }
            catch (Exception e)
            {
                Debug.WriteLine("MakeRequest - Exception message: " + e.Message);
                Debug.WriteLine("MakeRequest - Exception stack trace: " + e.StackTrace);
                Debug.WriteLine("MakeRequest - Exception HResult: " + e.HResult);
                Debug.WriteLine("MakeRequest - Exception inner ex: " + e.InnerException);
            }
        }


The exception:

Exception message: An error occurred while sending the request.
Exception stack trace:    at System.Net.Http.HttpClientHandler.<SendAsync>d__1.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter`1.GetResult()
   at Proj.Api.<MakeRequestExample>d__7.MoveNext()
Proj.Api=> Exception HResult: -2147012746
Proj.Api=> Exception inner ex: System.Runtime.InteropServices.COMException (0x80072F76): The text associated with this error code could not be found.

The requested header was not found

   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Net.Http.HttpHandlerToFilter.<SendAsync>d__1.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Net.Http.HttpClientHandler.<SendAsync>d__1.MoveNext()

**Comment by SidharthNabar** [comment]
@Hochman85 - It is not clear that the problem you are referring to is the same as the one tracked by this issue. Can you provide the following info:
1. Is there a redirect involved in your scenario? (this issue tracks the problem of a header set on the original request not being set on the subsequent request sent after a redirect)
2. In the code snippet above, where is the exception being thrown? 
3. As far as I can tell, your Catch block doesn't seem to get called - why is that?


I just tried your code snippet with my test server (no redirect) and the Client-Id header does show up on the server side. Is there a repro step I am missing?

**Comment by Hochman85** [comment]
@SidharthNabar Thanks for your reply Sid.

1. The problem is slightly different in a way that I am not using any redirect data before the request. There will eventually be a redirect after the request is sent as the server will return TemporaryRedirect. That is what I am expecting. But for now, I am not even able to send the request.
2. The exception is thrown at httpClient.SendAsync
3. My catch bloc is called and all those logs I printed were from the catch bloc. I am logging: exception message, stacktrace, HResult and inner exception - all included in my question.

To recap: the request works fine if I do not add the customer header (Client-Id). The moment I add that, I get the exception above. However, if I enable the line "Accept" (text/plain) commented out above I get NotAcceptable.

P.S. the same logic works in Windows Forms (similar project) but not in universal apps.

**Comment by davidsh** [comment]
@Hochman85 I see that you opened up a new issue #5065 about this.  So, we should move any further discussion of this issue to that new issue you opened up.

**Comment by Hochman85** [comment]
@davidsh Thanks David. Makes sense. I thought it could be not the same issue - that's why I opened a new one just in case.

**Comment by SidharthNabar** [comment]
@Hochman85 - Yep, it is not the same issue. Let's continue the discussion there.

**Comment by Hochman85** [comment]
@SidharthNabar No problem Sid. Looking forward to it :)
